nextroute/schema/fleet.go

```diff
@@ -0,0 +1,234 @@
+	MaxDistance             *int       `json:"max_distance,omitempty"`
+	MaxDuration             *int       `json:"max_duration,omitempty"`
+	StopDurationMultiplier  *float64   `json:"stop_duration_multiplier,omitempty"`
+	Backlog                 []string   `json:"backlog,omitempty"`
+	AlternateStops          []string   `json:"alternate_stops,omitempty"`
+	InitializationCost      int        `json:"initialization_cost,omitempty"`
+}
+
+// FleetStop holds the fleet input stop data.
+type FleetStop struct {
+	ID                      string       `json:"id,omitempty"`
+	Position                Location     `json:"position,omitempty"`
+	UnassignedPenalty       *int         `json:"unassigned_penalty,omitempty"`
+	Quantity                any          `json:"quantity,omitempty"`
+	Precedes                any          `json:"precedes,omitempty"`
+	Succeeds                any          `json:"succeeds,omitempty"`
+	HardWindow              *[]time.Time `json:"hard_window,omitempty"`
+	MaxWait                 *int         `json:"max_wait,omitempty"`
+	StopDuration            *int         `json:"stop_duration,omitempty"`
+	TargetTime              *time.Time   `json:"target_time,omitempty"`
+	EarlinessPenalty        *float64     `json:"earliness_penalty,omitempty"`
+	LatenessPenalty         *float64     `json:"lateness_penalty,omitempty"`
+	CompatibilityAttributes *[]string    `json:"compatibility_attributes,omitempty"`
+}
+
+// FleetToNextRoute takes a legacy cloud fleet input and converts it into
+// nextroute input format.
+func FleetToNextRoute(fleetInput FleetInput) (Input, error) {
+	input := Input{}
+	stopCompats := make([]string, 0)
+	vehicleCompats := make([]string, 0)
+	// Use default values and add special handling for CompatibilityAttributes
+	// and HardWindows.
+	if fleetInput.Defaults != nil && fleetInput.Defaults.Stops != nil {
+		input.Defaults = &Defaults{}
+		input.Defaults.Stops = &StopDefaults{
+			UnplannedPenalty:        fleetInput.Defaults.Stops.UnassignedPenalty,
+			Quantity:                fleetInput.Defaults.Stops.Quantity,
+			MaxWait:                 fleetInput.Defaults.Stops.MaxWait,
+			Duration:                fleetInput.Defaults.Stops.StopDuration,
+			TargetArrivalTime:       fleetInput.Defaults.Stops.TargetTime,
+			EarlyArrivalTimePenalty: fleetInput.Defaults.Stops.EarlinessPenalty,
+			LateArrivalTimePenalty:  fleetInput.Defaults.Stops.LatenessPenalty,
+		}
+		if fleetInput.Defaults.Stops.CompatibilityAttributes != nil {
+			stopCompats = *fleetInput.Defaults.Stops.CompatibilityAttributes
+		}
+		if fleetInput.Defaults.Stops.HardWindow != nil {
+			input.Defaults.Stops.StartTimeWindow = *fleetInput.Defaults.Stops.HardWindow
+		}
+	}
+
+	// Use default values. nil values are not compatible between fleet and
+	// nextroute.
+	if fleetInput.Defaults != nil && fleetInput.Defaults.Vehicles != nil {
+		vehicleCompats = fleetInput.Defaults.Vehicles.CompatibilityAttributes
+		input.Defaults.Vehicles = &VehicleDefaults{
```

Comment:
If the `defaults` for `stops` is nil, the above `if` condition will not be true, and thus `input.Defaults = &Defaults{}` will never be called and this line will probably panic with nil pointer dereference. I think.

Reply:
yep, fixed this locally. 👍